allow exclusion of resources using standard label

prydonius opened the pull request:
excludes any resources with the velero.io/exclude-from-backup=true label

fixes #1373 

Signed-off-by: Adnan Abdulhussein <aadnan@vmware.com>

prydonius commented:
I guess we should also add some documentation about this new feature, where do we think that should go?

prydonius commented on pkg/backup/backup_new_test.go:
```diff
@@ -218,6 +218,54 @@ func TestBackupResourceFiltering(t *testing.T) {
+				Backup(),
+			apiResources: []*apiResource{
+				pods(
+					withLabel(withLabel(newPod("foo", "bar"), "velero.io/exclude-from-backup", "true"), "a", "b"),
```
this was a lazy way of adding multiple labels to the pod, I can create another helper that takes multiple labels that should be easier to read

carlisia replied:
Not sure we have a place, but should go together with this: #1440

skriss replied:
Yeah, it would be nice to update the helper. One option that keeps it pretty concise would be to make it variadic, and just assume it goes `..., label key 1, label value 1, label key 2, label value 2, ...`. That way you don't have to use a map or anything else.

prydonius replied:
@skriss nice, I was thinking about something like that, will look into variadic variables

carlisia commented on pkg/backup/resource_backupper.go:
```diff
@@ -222,8 +222,10 @@ func (rb *defaultResourceBackupper) backupResource(group *metav1.APIResourceList
+		labelSelector = excludeLabelSelector
 		if selector := rb.backupRequest.Spec.LabelSelector; selector != nil {
-			labelSelector = metav1.FormatLabelSelector(selector)
+			labelSelector = labelSelector + "," + metav1.FormatLabelSelector(selector)
```
I would `s/labelSelector/"velero.io/exclude-from-backup!=true"` and delete lns 225,226. WDYT? I don't think this is so complex that needs to be expanded. Less lines is better if it's still clear.

skriss replied:
I'd recommend:

```go
labelSelector := "velero.io/exclude-from-backup!=true"
if selector := rb.backupRequest.Spec.LabelSelector; selector != nil {
    labelSelector = labelSelector + "," + metav1.FormatLabelSelector(selector)
...
```

prydonius replied:
Agree there's a few more lines than necessary here, I like @skriss's suggestion as it will make it easier to find just that labelSelector in the future

carlisia commented:
> Not sure we have a place, but should go together with this: #1440

This ^ was supposed to be in response to:

> I guess we should also add some documentation about this new feature, where do we think that should go?

skriss commented on pkg/backup/backup_new_test.go:
```diff
@@ -218,6 +218,54 @@ func TestBackupResourceFiltering(t *testing.T) {
+				"resources/pods/namespaces/zoo/raz.json",
+				"resources/persistentvolumes/cluster/bar.json",
+			},
+		},
```
Can you add a case that covers the new label key being specified, but the value not being true?

skriss commented:
Yeah, I'm not sure there's a great place to put docs for it right now -- agree it would make sense to go with https://github.com/heptio/velero/issues/1440, probably on a new page under "Use" in the TOC.

prydonius commented on pkg/backup/backup_new_test.go:
```diff
@@ -1795,12 +1870,14 @@ func newHarness(t *testing.T) *harness {
 	}
-	labels[key] = val
+	for i := 0; i < len(labelPairs); i += 2 {
+		labels[labelPairs[i]] = labelPairs[i+1]
```
@skriss this will panic if `labelPairs` is not even. I was thinking about catching that, but since the `testing.T` object isn't passed in here, we can't fail the test more gracefully. Not sure if we want to pass `t` into each `withLabel` call. We could also just panic with a more helpful error (e.g. "withLabel requires key-value pairs when specifying labels")

skriss replied:
I think panicking with a helpful error if there's an odd number of values should be fine. It's really just to detect programmer error.

prydonius commented:
> Yeah, I'm not sure there's a great place to put docs for it right now -- agree it would make sense to go with #1440, probably on a new page under "Use" in the TOC.

added a comment in #1440 to include this

Also updated based on review, PTAL!

prydonius commented:
@skriss really good, the existing tests are really helpful and pretty easy to work off of!
